### PR TITLE
Add interpreter for AllGatherOp

### LINKED ISSUE #1118: Add interpreter for AllGatherOp
Following the [testing guidelines](https://github.com/openxla/stablehlo/blob/main/docs/interpreter_checklist.md)

### PR DESCRIPTION
**Pull request by ghpvnist**
We have the following constraints in the spec:

```
(I1) `operand`: tensor or per-tensor quantized tensor.
(I2) `all_gather_dim`: constant of type `si64`.
(I3) `replica_groups`: 2-dimensional tensor constant of type `si64`.
(I4) `channel_id`: constant of type `si64`.
(I5) `use_global_device_ids`: constant of type `i1`.
(C1) `0 <= all_gather_dim < rank(operand)`.
(C2) `is_unique(replica_groups)`.
(C3) `size(replica_groups)` is defined as:
* `num_replicas` if `cross_replica` is used.
* `num_replicas` if `cross_replica_and_partition` is used.
* `num_processes` if `flattened_ids` is used.
(C4) `0 <= replica_groups < size(replica_groups)`.
(C5) If `use_global_device_ids = true`, then `channel_id > 0`.
(C6) `type(result) = type(operand)` except:
* `dim(result, all_gather_dim) =
  dim(operand, all_gather_dim) * dim(process_groups, 1)`.
```

These constraints will be comprehensively covered by the following tests:

```
I1: a) `operand` is not a tensor. (Covered by ODS).
I2: a) `all_gather_dim` is not a constant of type `si64`. (Covered by ODS).
I3: a) `replica_groups` is not a 2-dimensional tensor constant.
I3: b) `element_type(replica_groups) != si64`. (Covered by ODS).
I4: a) `channel_id` is not a constant of type `si64`. (Covered by ODS).
I5: a) `use_global_device_ids` is not a constant of type `i1`. (Covered by ODS).
C1: a) `all_gather_dim < 0`.
C1: b) `all_gather_dim >= rank(operand)`.
C2: a) `is_unique(replica_groups) = false`.
C3: a) `size(replica_groups)` is defined as:
* `num_replicas` if `cross_replica` is used.
* `num_replicas` if `cross_replica_and_partition` is used.
* `num_processes` if `flattened_ids` is used. 
C4: a) `replica_groups < 0`.
C4: b) `replica_groups >= size(replica_groups)`.
C5: a) `use_global_device_ids = true` and `channel_id <= 0`.
C6: a) `type(result) != type(operand)` except:
* `dim(result, all_gather_dim) =
  dim(operand, all_gather_dim) * dim(process_groups, 1)`.
```

If we drop the "Covered by ODS" pieces, this will leave us with the following test cases:

```
I3: a) `replica_groups` is not a 2-dimensional tensor constant.
C1: a) `all_gather_dim < 0`.
C1: b) `all_gather_dim >= rank(operand)`.
C2: a) `is_unique(replica_groups) = false`.
C3: a) `size(replica_groups)` is defined as:
* `num_replicas` if `cross_replica` is used.
* `num_replicas` if `cross_replica_and_partition` is used.
* `num_processes` if `flattened_ids` is used. 
C4: a) `replica_groups < 0`.
C4: b) `replica_groups >= size(replica_groups)`.
C5: a) `use_global_device_ids = true` and `channel_id <= 0`.
C6: a) `type(result) != type(operand)` except:
* `dim(result, all_gather_dim) =
  dim(operand, all_gather_dim) * dim(process_groups, 1)`.
```

Notes:
  * C3a verification is infeasible since `num_replicas` and `num_partitions` are not known statically.
  * C4b verification is infeasible since `size(replica_groups)` is not known statically.

closes #1118

### CODE REVIEW

File: docs/spec.md [hunk 1]
```diff
@@ -709,31 +709,32 @@ Afterwards, within each `process_group`:
...
 |----------|---------------------------------------|-------------|
-| `result` | tensor or per-tensor quantized tensor | (C6)        |
+| `result` | tensor or per-tensor quantized tensor | (C7)        |
 
 #### Constraints
 
-* (C1) `0 <= all_gather_dim < rank(operand)`.
-* (C2) `is_unique(replica_groups)`.
-* (C3) `size(replica_groups)` is defined as:
+* (C1) `dim(operand, all_gather_dim) > 0`.
```
Comment (burmako) [inline docs/spec.md:728]: I'm not sure where this constraint is coming from. Perhaps let's leave open a Specification Compliance ticket and leave this to another PR? Everything else LGTM.

File: stablehlo/dialect/StablehloOps.cpp [hunk 2]
```diff
@@ -877,9 +877,13 @@ void AllToAllOp::build(OpBuilder& odsBuilder, OperationState& odsState,
 //===----------------------------------------------------------------------===//
 
 LogicalResult AllGatherOp::verify() {
+  int64_t channelId = 0;
```
[the remark this answers is not in the document]
Comment (ghpvnist) [reply stablehlo/dialect/StablehloOps.cpp:880]: This file doesn't have access to `ChannelId` type, and probably shouldn't?

[the remark this answers is not in the document]
Comment (burmako) [reply stablehlo/dialect/StablehloOps.cpp:880]: You’re right! Nvm.

File: docs/spec.md (hunk 1 above)
Comment (ghpvnist) [reply docs/spec.md:728]: Created #1745

File: stablehlo/dialect/TypeInference.cpp [hunk 3]
```diff
@@ -3047,45 +3047,45 @@ LogicalResult verifyAllGatherOp(std::optional<Location> location, Value operand,
...
 
   if (operandType) {
-    // all_gather_c2
+    // all_gather_c1
     if (allGatherDim >= operandType.getRank())
       return emitOptionalError(
           location, "all_gather_dim must be a valid index of operand");
 
-    // all_gather_c1
+    // TODO()
```
Comment (burmako) [inline stablehlo/dialect/TypeInference.cpp:3060]: Needs a ticket number.

**Review by burmako — APPROVED** [review]

File: stablehlo/dialect/TypeInference.cpp (hunk 3 above)
Comment (ghpvnist) [reply stablehlo/dialect/TypeInference.cpp]: Done.
